### PR TITLE
Refactor&Feature: Mixing module

### LINKED ISSUE #2768: Reconstruction of the charge mixing module
### Describe Current Status and Possible Solution

The charge mixing codes will be reconstructed soon. Following aspects are considered this time:
1. correctness of mixing method name. In previous code the 'pulay' mixing_type is actually implemented as 'broyden', and 'broyden' is itself. This time this will be corrected.
2. implementation of standard Pulay mixing method.
3. feature support. For DFT+U needs mixing of density matrix, other quantities will be possible to mix in the future, while previous code only support mixing of charge density data, there is also another place (module_ri) where mixing of density matrix is implemented, this time the mixing of density matrix will be also supported, the related code in module_ri is also planned to re-organized.
4. heterogeneous programming. The container Tensor is designed for supporting heterogeneous computing. To minimize data exchange between devices, i.e., CPU and GPU, history data will be directly saved in Tensor in device perform computation.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Qianruipku**
In previous version "pulay" used simplified broyden method but is not the actual pulay method, so after this PR, all pulay should be replaced by broyden.
image

Fix #2768
Finish unit tests

### CODE REVIEW

**Comment by Satinelamp** [comment]
@Qianruipku, @dyzheng 
The failure in the `integration` test could be solved by merging the 'develop' branch into this PR.
Since in the PR(https://github.com/deepmodeling/abacus-develop/pull/2977), we have updated the reference value.
image

File: examples/deepks/lcao_CsPbI3/INPUT
Comment (kirk0830) [inline examples/deepks/lcao_CsPbI3/INPUT:1]: do we need to add `mixing_ndim 7` for all `lcao` calculations? if we want to keep "everything" identical.

File: source/Makefile.Objects
Comment (kirk0830) [inline source/Makefile.Objects:1]: terrific, seems you have adjusted codes in RI

File: source/module_base/module_mixing/broyden_mixing.h [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            free(F);
+        if (dF != nullptr)
+            free(dF);
+    };
+    virtual void reset() override
+    {
+        this->ndim_cal_dF = 0;
+        this->start_dF = -1;
+        this->address = nullptr;
+    }
```
Comment (kirk0830) [inline source/module_base/module_mixing/broyden_mixing.h:56]:
```suggestion
    }
/*
    After full implementation of Tensor container, this part will be refactored. Presently overload strategy is used.
*/
```
@denghuilu see if any related instruction can add here?

File: source/module_base/module_mixing/broyden_mixing.h [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+ *        beta{ij} = inv(alpha){ij}
+ *        c{mk} = <dF{k}, F{m}>
+ *        gamma{mn} = \sum_k c{mk} * beta{kn}
+ *        n{m+1} = n_in{m} + mixing_beta*F{m} - \sum_n gamma{mn} * (dn_in{n} + mixing_beta*dF{n})
+ *        mixing_data{i} = n_in{i} + mixing_beta*F{i}
+ *        n{m+1} = \sum_i coef{i} * mixing_data{i}
+ */
+class Broyden_Mixing : public Mixing
+{
+  public:
```
Comment (kirk0830) [inline source/module_base/module_mixing/broyden_mixing.h:31]: Maybe annotations are also needed here even if it is a header file of derived class? @dyzheng

File: source/module_base/module_mixing/test/mixing_test.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        double xnorm = 0.0;
+        for (int ir = 0; ir < 3; ++ir)
+        {
+            xnorm += x1[ir].real() * x2[ir].real() + x1[ir].imag() * x2[ir].imag();
+        }
+        return xnorm;
+    }
+};
+
+TEST_F(Mixing_Test, Broyden_Solve_LinearEq)
```
Comment (kirk0830) [inline source/module_base/module_mixing/test/mixing_test.cpp]:
```suggestion
TEST_F(MixingTest, BroydenSolveLinearEq)
```
underline "_" might be discouraged in unit test, if I remember correctly :(
If so, it is better to change names of test suite and tests, thanks

File: source/module_esolver/esolver_ks.cpp [hunk 4]
```diff
@@ -84,6 +70,23 @@ namespace ModuleESolver
     void ESolver_KS<T, Device>::Init(Input& inp, UnitCell& ucell)
     {
         ESolver_FP::Init(inp,ucell);
+
+        //------------------Charge Mixing------------------
+        p_chgmix->set_mixing(INPUT.mixing_mode,
+                             INPUT.mixing_beta,
+                             INPUT.mixing_ndim,
+                             INPUT.mixing_gg0,
+                             INPUT.mixing_tau);
```
Comment (kirk0830) [inline source/module_esolver/esolver_ks.cpp]: I am not sure here about whether it is appropriate to directly pass variable from Input class to inner class the Charge_Mixing, will GlobalV be a better choice? I mean from Input to GlobalV, then pass variables between GlobalV and Charge_Mixing. @dyzheng

**Review by kirk0830 — COMMENTED** [review]
I add several comments. But overall there is a more important thing, it is, whether this PR is a refactor, or because it also brings new feature so it should be also a "Feature". If so, I guess more tests would be required...

File: source/module_base/module_mixing/test/mixing_test.cpp (hunk 3 above)
Comment (Qianruipku) [reply source/module_base/module_mixing/test/mixing_test.cpp]: Ok

File: examples/deepks/lcao_CsPbI3/INPUT
Comment (Qianruipku) [reply examples/deepks/lcao_CsPbI3/INPUT:1]: mixing_ndim 8 is better than mixing_ndim 7. We do not need to make the energy of each iteration same. We only need to keep the final results same. Mixing_ndim does not change final results.

File: source/module_esolver/esolver_ks.cpp (hunk 4 above)
Comment (dyzheng) [reply source/module_esolver/esolver_ks.cpp]: Values of class INPUT should not used in other codes, only input_conv.cpp can access these variables.

**Comment by dyzheng** [comment]
I have no problem of the Refactor part for this PR, but the Feature part not related to the Issue.

**Review by dyzheng — APPROVED** [review]

**Review by kirk0830 — APPROVED** [review]
Cant wait to see good performance brought about by this PR and on-coming our fully implementation of new DFT+U code

**Review by kirk0830 — APPROVED** [review]
these changes are due to conflicts resolve.

**Comment by hongriTianqi** [comment]
This PR also solved #3040

**Pull request merged**